### DISCUSSION TITLE
Adding morphisms to Pint.

### DISCUSSION THREAD

**Discussion by otrho** [essential-contributions/pint#576] [Ideas]
Issue #133 is about adding Pint syntax for a more general `forall` and for `map`, `fold`, `filter`, etc.  I have a bunch of questions which need research.

1. Is it possible to compile `map` or `fold` bodies using _only_ a data stack and a `rep` loop?
2. What do we do with temporary arrays, e.g., the result of `map` or `filter`?
3. Can we support arbitrary depth for nested morphisms?

## Simple Stack VM

So far I have an answer for (1), which is yes.  But it's quite limited.  I wanted to know how do you even compile an expression like `map(ary, |el| { 1 + el * 2 + el + el + el })`, or `fold(ary, 0, |acc, el| { acc * 10 + 100 - el })` with only a stack and the answer I came up with is in a [PoC](https://github.com/otrho/morphism-poc).

Usually in codegen, or compiling overall, one will come up with a general solution which works for all cases but may be way too inefficient.  But the general case is always a backup which can then be optimised.

For the PoC I found that this isn't really possible at the basic level, and every context needed to be a special case, where a 'context' is whether an expression is outside of a morphism, is it in a `map` body with a single arg or in a `fold` body with two args, and even the specific order by which the args are accessed.

Not general at all.  So I think having just a data stack for this stuff isn't enough.

## Temporaries

Ideally temporaries go on the stack and I think 90% this will be obvious and easy.  The nature of a constraint based model is that everything eventually boils down to a boolean value.  The result of a `map` or `filter` must be either compared to a solution or be reduced down into an arithmetic or logical conjunction.

Having the temporary value on the stack should be manageable, complicated only when we need to make multiple distinct reductions of the same array.  Perhaps copying arrays on the stack is the answer, perhaps stashing them to a memory is the answer.  The former isn't great for concurrency, the latter complicates the VM environment.

## Generalised Nesting

We already have nesting in the Pint generators `forall` and `exists`.

Ideally we'd replace the generators with folds of maps, e.g., `forall x in ary { x > 0 }` becomes `fold(map(ary, |x| { x > 0 }), true, |a, x| { a && x })`, which in turn can of course be deforested to `fold(ary, true, |a, x| { a && x > 0 })`.  I say ideally, but that's my opinion I guess -- due to avoiding control flow in Pint's output we unroll all the generators, which is pretty bloaty.  If we're going to allow loops then we might as well use them and avoid the bloat.

But `forall` allows arbitrary nesting, e.g., `forall i in 0..10, j in 0..5, k in ary { ... using i, j, k ...}` and this is not really feasible as `map(0..10, |i| { map(0..5, |j| { map ary, |k| { ... i, j, k ... }))` with purely stack operations.

I _have_ found that having two variables is possible, as with `fold`, and a `map` body can use the same mechanisms for evaluation as a `fold` body by having a dummy accumulator, albeit less efficient.  And so we could decide the general case for all morphism bodies is a two parameter lambda expression and no more.  I.e., nesting beyond a double `map` is illegal.

This seems a bit crap to me though.

## A 'Register Based VM'

I have always thought a register based VM meant a VM resembling a modern CPU, which is exactly what the Fuel VM was.  But this is not quite accurate and in the scheme of things doesn't make sense.  A CPU has a limited set of registers for efficiency's sake -- a CPU register is the fastest storage possible, faster than cache or RAM -- but a VM doesn't have this advantage as any storage, whether it's pretending to be a CPU register or not, is all just regular memory.

So a register VM can provide an arbitrary number of 'registers' to the execution environment at no more or less cost.

Lua was the first VM to do this.  I'm not 100% sure, but I think Lua instructions provide a 9-bit register index, which implies that there may be up to 512 addressable registers available, and each function gets a fresh context, registers aren't necessarily shared and/or need to be preserved.  I think.  But Lua only has instructions for manipulating these instructions, i.e, `ADD` will always add two registers and store the result in another register.

Wasm has chosen a hybrid approach which I'm quite fond of, and I think we should adopt something similar.  Arithmetic operations are still stack based, i.e., `ADD 1 2` is actually `PUSH 1`, `PUSH 2`, `ADD`, but Wasm functions have indexed read/write locals which may be referenced from anywhere.  This provides both the benefit of a simple instruction set -- especially decoding -- that stack ops provide along with the mechanism for carrying arbitrary context along with evaluation.

So the nesting problem of having to manage a number of parameters, `i`, `j`, `k` in a lambda body can be simplified by using directly accessed local 'registers'.

Specifically, I'm thinking the control stack can also keep a data stack index to the beginning of the current context, i.e., where the local registers live.  Then we provide instructions to get or set these values by index based on that context base.  It's not dissimilar to how `BP` is used for x86 binaries, marking on the stack where locals or function arguments are.

And these get/set instructions should be able to reference prior contexts, so `k` might be in the current context but `j` in the immediately prior and `i` in the one before that.

## So...

So... that's what I've been thinking about and have decided to open this discussion.

**Comment by freesig** [comment]
> The former isn't great for concurrency
What does this mean? We don't have any concurrency within a single constraint.

**Reply by otrho** [reply]
If we consider that constraint boundaries aren't necessarily concrete, and if we were to optimise for concurrency we may want to join or split boolean expressions.  If an array is used in multiple separate boolean expressions and is in memory then they may be evaluated concurrently.  If the array is on the stack then this is impossible.

**Reply by freesig** [reply]
Yeh I think this is when it's better to do:
```
let arr: int[len] = ${some_temp_array_exp};
```
Then the solver has to provide `arr` upfront so it can be used in multiple constraints without creating any contention.

**Reply by otrho** [reply]
Yeah, it depends on how `arr` is used.  All arrays are eventually boiled down to bools somehow.  Maybe only `arr[0]` is actually used, the rest can be thrown away.  Maybe it's hashed or summed or some other fold.  Again it doesn't necessarily need to go into memory.

So only if multiple parts are needed in separate constraints do we need to stash things away, and not necessarily the whole array.

**Reply by freesig** [reply]
Note that this isn't going into memory. The solver has to provide `arr` so it's a decision variable.

**Reply by otrho** [reply]
Yeah, I was thinking in a perf context, to avoid performing the morphism multiple times in different constraints.  The more complexity we add to the VM the harder it becomes for solvers.

Like if we have an enormous nested morphism building a giant multidimensional array and pluck a single element out and/or fold it all down to a scalar, then there's one decision variable needed.  And so the solver solves it, but to do so needs to understand what the giant array represents.  Tricky.

**Reply by freesig** [reply]
You would only perform the morphism once though because the solver is just providing the a dec var array that matches the morphisms result

**Reply by otrho** [reply]
True, true, the other constraints don't need to redo it -- as a set they all need to pass and so doing it once is enough -- no need to prove the same thing twice.

In that case the model is actually much simpler (well, it was always simpler, it's just becoming simpler in my mind) and the compiler can know exactly how many temporaries it needs for any particular constraint, and they can live in global memory just as easily as locals.

If memory is essentially local to the constraint and therefore temporary then I think it's simpler.  The 1000 word example from above would never be necessary... well, it could only be necessary with 500 _nested_ folds.  The compiler knows that only a single non-nested morphism runs at a time so the temporaries could be re-used.

I think your concern about allocation cost is a little unfounded.  Each constraint, or computation unit, or whatever can pre-declare the size of its working memory.  The VM can then guarantee it doesn't stray with loads and stores outside of that boundary.  All the constraints in an intent can have their memory sizes summed and only a single allocation is then needed per intent.  Each constraint then gets its own slice.  Does that work?

**Reply by freesig** [reply]
How would a constraint know how much memory to pre declare?

**Reply by otrho** [reply]
I was thinking about this some more this morning in the shower, where I do all my serious thinking.  I'm sure that 90% of the time the compiler wouldn't need to store arrays in memory and working on the stack would be enough.

Nested morphisms should be deforested into single passes.  Folds can just be done in place, including comparisons with ranges of decision vars.  Therefore it's only temporaries for the 'args' to the morphisms which need to go in memory, which are known at compile time.  At most 2 words per go, multiplied by the max nest depth == total memory for temporaries.

But yeah, there's a <10% chance that a temporary array can't stay on the stack and needs to go into memory, at which time the size could be unknown.  I honestly can't think of a situation where it couldn't be avoided off hand, but the mechanisms to do so need to be available to the compiler.

Like, comparing two filtered arrays.  The simplest (most naive?) basic method, which is guaranteed to be correct is to put both arrays in memory and then compare them element-wise afterwards.  But it would be possible to perform the filters simultaneously and compare as you go.  This would be really tricky though and involve some very special canoodling in the compiler which we would have to provide.  But it's _possible_ to avoid runtime memory allocations.

An alternative strategy in that case might be to just put the filtered arrays into decision variables and split the comparison into 3 different constraints; one each for the filters and another for the comparison.  But I'm thinking we generally want to avoid creating decision variables purely to act as temporaries.  Our aim should be to reduce the decision variable space to those only needed directly by the state transition IMO.  Perhaps that's overly strict?  Surely temp vars is a bad idea though.

**Comment by freesig** [comment]
I'm not really sure I see the difference between wasms locals and memory? Isn't that what it is? Or is there some sense of scoping so that locals are dropped when they go out of scope?

**Reply by otrho** [reply]
They're on the data stack so yes, they go out of scope.  If they were in memory then each morphism would need dedicated space for the variables.  This is fine and perhaps worth considering if we expect the average program to have only a few of these morphisms.

Consider a program with 5 uses of `fold`.  Each one needs somewhere to store the accumulator and current element, so that's two words x5 = 10 words of memory.  Each pair is used only while their particular `fold` is executing.   This sounds find, 10 words isn't much.

Using locals there would be no use of memory since the accumulator and element are stored on the data stack.  As each fold ends those values are popped off.  They take up at most 2 words unless they are nested.

So both versions seem OK, have fairly similar complexity.  But if we have a program with 500 folds then the memory use becomes 1000 words (8KB?) for the former and still only 2 words for the latter.

Then again, it also depends on our memory model for globals.  Are they shared between constraints and persist the entire time or does each one get its own memory address space?  If it's the latter then those 1000 words aren't necessarily all allocated at once.

**Reply by freesig** [reply]
I like the locals model a lot better because it keeps the maximum amount of memory usage lower and avoids a lot of allocations. If we did have global memory it would only be scoped within a single constraint. If we allowed mutable memory to exist across multiple constraints we'd either loose the ability to run them in parallel or have to introduce some form of locking around the memory. Both options are pretty bad so I think it's better we keep everything scoped within a single constraint. Also constraints should be able to be executed independently to allow them to be sampled. If they depend on other constraints executing that would break this.

I think if we can make all this work with a register of locals that pop off when they go out of scope that would be ideal from the POV of executing the VM.

**Reply by otrho** [reply]
Hang on then.  If we have the result of a morphism going at least partially into memory so that they can be shared between constraints as discussed [above](https://github.com/essential-contributions/pint/discussions/576#discussioncomment-9361260), where are they going?

And so I think all that I'm proposing for locals is the ability to declare a new 'scope' which puts the size of the current data stack somewhere handy, like on the control stack, and then for instructions `local_get idx` and `local_set idx val` to be added, which base the `idx` arg against the control stack scope base.  The actual pushing and popping would still be manually done by the program/compiler.  If the `idx` is out of bounds then we simply halt.  And so `enter` followed immediately by `local_get` would halt since nothing's there to get...

Except yeah, parent scopes should be available, so `local_get scope idx` might be needed, where `scope` is 0 for the current locals, 1 for the parent.  But we wouldn't need `local_set` for parent scopes.

**Reply by freesig** [reply]
Yeh I don't think we can share the result of a morphism between constraints. We can only compare it to a decision variable and then share the decision variable between constraints.

**Reply by otrho** [reply]
Fair enough.
